### ISSUE TITLE
Ping hangs in github.com/go-hep/csvutil/driver_test.go

### ISSUE THREAD

**Issue by cznic** [cznic/ql#149]
[Here](https://github.com/go-hep/csvutil/blob/e6468a13070ff5d437941213a0b2467bcd6c256a/csvdriver/driver_test.go#L36).

**Comment by sbinet** [comment]
FYI, if I put the `Ping` outside of the transaction, it's not hanging anymore:

```diff
diff --git a/csvdriver/driver_test.go b/csvdriver/driver_test.go
index 539bfc0..b7110ee 100644
--- a/csvdriver/driver_test.go
+++ b/csvdriver/driver_test.go
@@ -26,6 +26,11 @@ func TestOpen(t *testing.T) {
        }
        defer db.Close()
 
+       err = db.Ping()
+       if err != nil {
+               t.Fatalf("error pinging db: %v\n", err)
+       }
+
        tx, err := db.Begin()
        if err != nil {
                t.Errorf("error starting tx: %v\n", err)
@@ -33,13 +38,6 @@ func TestOpen(t *testing.T) {
        }
        defer tx.Commit()
 
-       /*
-               err = db.Ping()
-               if err != nil {
-                       t.Fatalf("error pinging db: %v\n", err)
-               }
-       */
-
        rows, err := tx.Query("select var1, var2, var3 from csv order by id();")
        if err != nil {
                t.Errorf("error querying db: %v\n", err)
```

not sure whether this is by design (_ie_ me not properly reading the docs of `"ql"` and/or `"database/sql"`) or a bug :)

**Comment by cznic** [comment]
I think it's a bug in ql, it triggers something like a livelock in the transaction state machine, but that's all I know from the first  two minute look (issuing SIGQUIT and seeing the first stack trace).

**Comment by cznic** [comment]
@sbinet 

I think I've fixed the livelock issue. However, that does not solve the problem.

Initiaiting transactions concurrently is fine. R/O transactions (basically only `SELECT` statements) are not a problem, they just execute concurrently. R/W transactions must be isolated from any other transaction (R/O or R/W) and this is done by serialization. Simply said, the R/W transaction first waits for any other transaction to finish and then executes exclusively. It's the same way how `sync.RWMutex` is handling things.

There's a catch in the above. If a R/W transaction is started and it is not yet finished (`COMMIT`) and the same goroutine attempts to start another R/W transaction in the same DB *not concurrently*, the serialization mechanism unavoidably(*) causes a deadlock.

The question remaining is why `sql.Ping` initiates a R/W transaction. That might still be my fault, but I haven't yet traced the source of the transaction.

----

*: As far as I can tell, ie. I don't see any solution to this problem. Any ideas welcome ;-)

**Comment by cznic** [comment]
@sbinet 

I think I finally understand the problem.

[`db.Ping:137`](https://github.com/go-hep/csvutil/blob/e6468a13070ff5d437941213a0b2467bcd6c256a/csvdriver/driver_test.go#L37) calls [`slq.Ping:499`](https://github.com/golang/go/blob/6b36535cf382bce845dd2d272276e7ba350b0c6b/src/database/sql/sql.go#L499) calls [`db.conn:780`](https://github.com/golang/go/blob/6b36535cf382bce845dd2d272276e7ba350b0c6b/src/database/sql/sql.go#L780) calls [`csvdriver.Open:100`](https://github.com/go-hep/csvutil/blob/e6468a13070ff5d437941213a0b2467bcd6c256a/csvdriver/driver.go#L100) calls [`csvdriver.importCSV:18`](https://github.com/go-hep/csvutil/blob/e6468a13070ff5d437941213a0b2467bcd6c256a/csvdriver/import.go#L18) which executes

[`32: tx, err := conn.Begin()`](https://github.com/go-hep/csvutil/blob/e6468a13070ff5d437941213a0b2467bcd6c256a/csvdriver/import.go#L32).

IOW, a `database/sql` driver cannot initiate a R/W transaction in its `Open` method without deadlocking any database implementing transaction isolation by serialization (SQLite does the same IIRC) because `driver.Open` can be invoked unpredictably by `database/sql` to, for example, create a pool of 'spare' connections, or, as in this case, just for pinging the database/driver.

This is caused by `database/sql` reusing the DB DSN when openning a new connection to the DB for `Ping`. `csvdriver` interprets the same DSN by attempting to once again create the same DB from the same data, causing a nested R/W TX on a different connection which is quire probably not correct. At least it causes the deadlock discussed above.

I suggest to adopt something like the mechanism used in [`ql/driver.go:195`](https://github.com/cznic/ql/blob/7a63cd7aa46ecd9f549b96983029576af2178f60/driver.go#L195). It manages a map of DSNs, opening an already opened DSN reuses the internal db instance. Lifetimes of shared internal db instances are controlled easily by updating refcounts as approprite.

I'm now gonna commit the livelock fix, but go-hep/csvutil#3 has to be fixed in `csvdriver` separately.
